lib/ansible/plugins/strategy/free.py

```diff
@@ -180,6 +188,7 @@ def run(self, iterator, play_context):
                                 del self._blocked_hosts[host_name]
                                 continue
 
+                        task_count += 1
```

Comment:
I believe counting tasks is not sufficient. Say there are 3 hosts and 2 forks.

At the top of the inner loop, first iteration: the first host is still executing a long running regular non-meta task from previous iteration and therefore is blocked and skipped: still `workers_free == 1`.

Second iteration, the second host is about to execute a meta task and does so: `workers_free == 1`, `task_count == 1`.

With the state of the counters above the inner loop continues to the third iteration and the third host can queue/execute a task (either regular or meta task) and only then the loop would end (either via `workers_free == 0` or `task_count == 2 == fork_max`... and the fact that there are no more hosts left but we wouldn't get to that check) resulting in 3 hosts executing on 2 forks 💥

Carefully crafted playbook to demonstrate:

```yaml
- hosts: localhost0,localhost1,localhost2
  gather_facts: false
  strategy: host_pinned
  tasks:
    - debug:
        msg: start

    - command: sleep 3
      when: inventory_hostname == "localhost0"

    # e.g. role_complete via include_role
    - meta: noop

    # e.g. flush_handlers
    - meta: noop

    - debug:
        msg: end
```

https://github.com/ansible/ansible/pull/83437 and https://github.com/ansible/ansible/pull/83438 handle this scenario.